ergo-installer.sh

```diff
@@ -0,0 +1,200 @@
+  echo 'ERROR: "--api-key" is required and requires a non-empty option argument.'
+  exit 1
+fi
+
+if [ "${MODE}" != 'full' ] && [ "${MODE}" != 'mining' ]; then
+  echo 'ERROR: "--mode" requires a non-empty option argument that can be "full" or "mining" only.'
+  exit 1
+fi
+
+echo "Ergo node with config file will be installed into '${APP_DIR}' directory and will be named as '${NODE_NAME}' and has API key '${API_KEY}'."
+
+
+## Check preinstalled software
+INSTALLATION_RECOMMENDATIONS=
+which java > /dev/null
+if [ $? != 0 ]; then
+  INSTALLATION_RECOMMENDATIONS="${INSTALLATION_RECOMMENDATIONS}\nsudo apt install default-jre-headless"
+fi
+
+which curl > /dev/null
+if [ $? != 0 ]; then
+  INSTALLATION_RECOMMENDATIONS="${INSTALLATION_RECOMMENDATIONS}\nsudo apt install curl"
+fi
+
+which daemon > /dev/null
+if [ $? != 0 ]; then
+  INSTALLATION_RECOMMENDATIONS="${INSTALLATION_RECOMMENDATIONS}\nsudo apt install daemon"
+fi
+
+if [ "${TOR}" != "no" ]; then
+  which tor > /dev/null
+  if [ $? != 0 ]; then
+    INSTALLATION_RECOMMENDATIONS="${INSTALLATION_RECOMMENDATIONS}\nsudo apt install tor"
+  fi
+  NODE_PARAMS="${NODE_PARAMS} -DsocksProxyHost=localhost -DsocksProxyPort=9050"
+  CONFIG_TEMPLATE="${CONFIG_TEMPLATE}\nscorex.network.bindAddress = \"127.0.0.1:9030\"\nscorex.restApi.bindAddress = \"127.0.0.1:9053\""
+fi
+
+if [ -n "${INSTALLATION_RECOMMENDATIONS}" ]; then
+  echo "To run Ergo node as you desired, you first need to install required software. For Ubuntu, run:${INSTALLATION_RECOMMENDATIONS}"
+  echo "After software installed, just run this script again with same parameters"
+  exit 1
+fi
+
+curl --output /dev/null --silent --head --fail http://localhost:9053
+if [ "$?" = 0 ]; then
+  echo "Seems node is already running on your machine (port 9053 is occupied), abort!"
+  exit 1
+fi
+
+# Create APP_DIR
+if [ -d ${APP_DIR} ]; then
+  echo "Ergo directory '${APP_DIR}' already exists."
+else
+  mkdir -p ${APP_DIR}
+  echo "Ergo directory '${APP_DIR}' created."
+fi
+
+# Download jar files from github
+LATEST_ERGO_RELEASE=$(curl --silent "https://api.github.com/repos/ergoplatform/ergo/releases/latest" | grep -Po '"tag_name": "\K.*?(?=")')
+LATEST_ERGO_RELEASE_NUMBERS=$(echo ${LATEST_ERGO_RELEASE} | cut -c 2-)
+ERGO_DOWNLOAD_URL=https://github.com/ergoplatform/ergo/releases/download/${LATEST_ERGO_RELEASE}/ergo-${LATEST_ERGO_RELEASE_NUMBERS}.jar
+echo "Latest known Ergo release: ${LATEST_ERGO_RELEASE}, downloading it to ${APP_DIR}/ergo.jar with overwrite..."
+curl --silent -L ${ERGO_DOWNLOAD_URL} --output ${APP_DIR}/ergo.jar
+echo "Ergo was downloaded to ${APP_DIR}/ergo.jar"
+
+# First run of Ergo node, to create API key hash that later be added into config
+echo -n "Executing ergo node to obtain API key hash..."
+daemon --chdir=${APP_DIR} -- java -jar ${APP_DIR}/ergo.jar --mainnet
+PID=$(pgrep -f "daemon --chdir=${APP_DIR} -- java")
+while ! curl --output /dev/null --silent --head --fail http://localhost:9053; do sleep 1 && echo -n '.'; done;
```

Comment:
@andyceo please comment on what happens in this loop and the other loops

Reply:
Done